runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -2625,3 +2625,188 @@ J9::TransformUtil::refineMethodHandleLinkTo(TR::Compilation* comp, TR::TreeTop*
+   //     can be calculated by simply adding ref and offset.
+   // 2.) The type is known to be an array at compile time. In this scenario, if the ref at runtime is `null` then
+   //     final address is calculated as in case 1. If not `null` then final address is the adjusted ref, by loading
+   //     the dataAddr pointer field then add it to the offset.
+   // 3.) The type of the object at dest is unknown at compile time (type is `java/lang/Object` or an interface).
+   //     In this scenario, a runtime null check and array check must be generated to determine whether it needs to 
+   //     be handled such as case 1 or case 2.
+   //     Interface is included as valid bytecode can store an array into an interface type and then gets passed to
+   //     Unsafe.copyMemory(). 
+
+   TR::Node *src        = arraycopyNode->getChild(1);
+   TR::Node *dest       = arraycopyNode->getChild(3);
+
+   // Check src/dest type at compile time
+   int srcSigLen, destSigLen;
+   const char *srcObjTypeSig = src->getSymbolReference() ? src->getSymbolReference()->getTypeSignature(srcSigLen) : 0;
+   const char *destObjTypeSig = dest->getSymbolReference() ? dest->getSymbolReference()->getTypeSignature(destSigLen) : 0;
+
+   // Case 3
+   bool srcArrayCheckNeeded, destArrayCheckNeeded;
+   if (!srcObjTypeSig)
+      srcArrayCheckNeeded = true;
+   else
+      {
+      TR_OpaqueClassBlock *srcClass = comp->fej9()->getClassFromSignature(srcObjTypeSig, srcSigLen, src->getSymbolReference()->getOwningMethod(comp));
+      srcArrayCheckNeeded = srcClass == NULL ||
+                              srcClass == comp->getObjectClassPointer() ||
+                              TR::Compiler->cls.isInterfaceClass(comp, srcClass);
+      }
+
+   if (!destObjTypeSig)
+      destArrayCheckNeeded = true;
+   else
+      {
+      TR_OpaqueClassBlock *destClass = comp->fej9()->getClassFromSignature(destObjTypeSig, destSigLen, dest->getSymbolReference()->getOwningMethod(comp));
+      destArrayCheckNeeded = destClass == NULL ||
+                              destClass == comp->getObjectClassPointer() ||
+                              TR::Compiler->cls.isInterfaceClass(comp, destClass);
+      }
+
+   // Case 2 & 3
+   bool srcAdjustmentNeeded = srcArrayCheckNeeded || srcObjTypeSig[0] == '[';
+   bool destAdjustmentNeeded = destArrayCheckNeeded || destObjTypeSig[0] == '[';
+
+   TR::TransformUtil::separateNullCheck(comp, arrayCopyTT);
+
+   if (!(srcAdjustmentNeeded || destAdjustmentNeeded))
+      {
+      TR::TransformUtil::convertUnsafeCopyMemoryCallToArrayCopyWithSymRefLoad(comp, arrayCopyTT, NULL, NULL);
+      return;
+      }
+
+   // Anchor nodes
+   for (int32_t i=1; i < arraycopyNode->getNumChildren(); i++)
+      {
+      TR::Node* childNode = arraycopyNode->getChild(i);
+      if ( !(childNode->getOpCode().isLoadConst() || childNode->getSymbolReference()->getSymbol()->isAutoOrParm()) )
```

Comment:
One more thing :sweat_smile:

Before calling `childNode->getSymbolReference()`, this should check to see that `childNode` is the kind of operation we're looking for:
```
childNode->getOpCode().isLoadVarDirect() && childNode->getSymbolReference()->getSymbol()->isAutoOrParm()
```